### PR TITLE
21061 add defaults to zoom preload

### PR DESCRIPTION
**Pull request by kriswest**
feat: #id [21061]

[Kanban link](https://chartiq.kanbanize.com/ctrl_board/48/cards/21061/details)

**Description of change**
* Add support for default zoom config setting at `foreign.components.['Window Manager'].zoomDefault` in the component config

**Description of testing**
1. Edit the Welcome Component config and set `foreign.components.['Window Manager'].zoomDefault = 0.5`
1. Spawn the Welcome component and check that it is zoomed out to 50%
1. change the zoom level and save into a workspace
1. reload the workspace and ensure that the zoom level is preserved in state in preference to default zoom level

### CODE REVIEW

File: src-built-in/preloads/zoom.js [hunk 1]
```diff
@@ -191,20 +199,33 @@ const zoomConfigHandler = (err, zoom) => {
...
+	} else if (zoomLevel != null) {
+		FSBL.Clients.Logger.info(`Retrieved zoomLevel from state: ${zoomLevel}`, err);
 		window.fsblZoomLevel = zoomLevel;
 		setZoom(window.fsblZoomLevel);
+		window.settingInitialZoom = false;
+	} else {
+		//check for default configuration for zoom level and apply as needed
+		let defaultLevel = _.get(FSBL.Clients.WindowClient.options.customData, "foreign.components.['Window Manager'].zoomDefault");
+		if (defaultLevel) { 
+			FSBL.Clients.Logger.info(`Retrieved default zoom level from config: ${defaultLevel}`, err);
```
Comment (NinjaManatee) [inline src-built-in/preloads/zoom.js]: I don't think `err` should be an argument here.

File: src-built-in/preloads/zoom.js [hunk 2]
```diff
@@ -77,19 +79,25 @@ const setZoom = (pct) => {
...
+	return Math.round((input + Number.EPSILON) * 10) / 10 ;
+}
+
 /**
  * Zooms the page in one step.
  */
 const zoomIn = () => {
-	window.fsblZoomLevel += window.zoomStep;
+	window.fsblZoomLevel = roundTo1Decimal(window.fsblZoomLevel + window.zoomStep);
+	if  (window.fsblZoomLevel < window.zoomMin) {window.fsblZoomLevel = window.zoomMin;}
```
[the remark this answers is not in the document]
Comment (kriswest) [reply src-built-in/preloads/zoom.js]: I'm aware - I added it here as well as in setZoom you can go past the range and the value applied is corrected back to, but you are still going past. If you scroll back the other way you have to make it back into the range before you start seeing the effect.  This way if reverse direction it is applied immediately. Arguably a minor nit, but did feel odd.

Added the rounding after ending up at 99% zoom a few times.

File: src-built-in/preloads/zoom.js [hunk 3]
```diff
@@ -77,19 +79,25 @@ const setZoom = (pct) => {
...
 	setZoom(window.fsblZoomLevel);
 }
 
 /**
  * Zooms the page out one step.
  */
 const zoomOut = () => {
-	window.fsblZoomLevel -= window.zoomStep;
+	window.fsblZoomLevel = roundTo1Decimal(window.fsblZoomLevel - window.zoomStep);
+	if  (window.fsblZoomLevel > window.zoomMax) {window.fsblZoomLevel = window.zoomMax;}
```
[the remark this answers is not in the document]
Comment (kriswest) [reply src-built-in/preloads/zoom.js]: As above, there is a (minor) reason to do it here as well. Could be moved to setZoom, which would have to update window.fsblZoomLevel (to stop you scrolling past the allowed range)

File: src-built-in/preloads/zoom.js [hunk 4]
```diff
@@ -191,20 +199,33 @@ const zoomConfigHandler = (err, zoom) => {
...
 /**
- * Applies the zoom level from the component state.
+ * Applies the zoom level from the component state or default configuration.
  *
  * @param {*} err The error, if one occurred, from getting the zoom level from component state.
  * @param {Number} zoomLevel The zoom level saved in the component.
  */
 const getZoomLevelHandler = (err, zoomLevel) => {
 	if (err) {
 		FSBL.Clients.Logger.info("No \"fsbl-zoom\" settings found in component state", err);
```
[the remark this answers is not in the document]
Comment (kriswest) [reply src-built-in/preloads/zoom.js]: It was moved inside the block in the first version of these changes, where the config was retrieved asynchronously, so it couldn't be set outside the block (and apparently missed the error state). Switched to a synchronous style and forgot to put it back, fixed.

File: src-built-in/preloads/zoom.js [hunk 5]
```diff
@@ -191,20 +199,33 @@ const zoomConfigHandler = (err, zoom) => {
...
-	}else if (zoomLevel != null) {
+	} else if (zoomLevel != null) {
+		FSBL.Clients.Logger.info(`Retrieved zoomLevel from state: ${zoomLevel}`, err);
```
[the remark this answers is not in the document]
Comment (kriswest) [reply src-built-in/preloads/zoom.js]: clearly I was half asleep.

File: src-built-in/preloads/zoom.js (hunk 1 above)
Comment (kriswest) [reply src-built-in/preloads/zoom.js]: fixed

File: src-built-in/preloads/zoom.js [hunk 6]
```diff
@@ -191,20 +199,33 @@ const zoomConfigHandler = (err, zoom) => {
...
+	} else {
+		//check for default configuration for zoom level and apply as needed
+		let defaultLevel = _.get(FSBL.Clients.WindowClient.options.customData, "foreign.components.['Window Manager'].zoomDefault");
+		if (defaultLevel) { 
+			FSBL.Clients.Logger.info(`Retrieved default zoom level from config: ${defaultLevel}`, err);
+			window.fsblZoomLevel = defaultLevel;
+			setZoom(defaultLevel); 
+			window.settingInitialZoom = false;
+		} else {
+			FSBL.Clients.Logger.info("No default zoom level retrieved from config: ",response.data, err);
```
[the remark this answers is not in the document]
Comment (kriswest) [reply src-built-in/preloads/zoom.js]: fixed

**Comment by kriswest** [comment]
Resolved a console warning caused by a passive event handler for the mouse wheel.

**Comment by kriswest** [comment]
> Using lodash, but not requiring it in.

Fixed.

Have not changed the min/max handling in zoomIn/zoomOut because I'd have to drop the argument from setZoom and have it read/control the value of window.fsblZoomLevel itself to stop the scroll heading on past the max/min

**Comment by kriswest** [comment]
Target changed

**Comment by kriswest** [comment]
@JMCCROWELL Where should docs for the zoom preload and its configuration go? Should they get entries in the config reference? The other place they could go is the [sample preloads section](https://documentation.chartiq.com/finsemble/tutorial-integratingHTMLApplications.html#zoom-js)
Existing config from zoom.js is global: `finsemble.Window Manager.zoom.timeout` / `step` / `min` / `max`
See [code comments](https://github.com/ChartIQ/finsemble-seed/blob/517d4a1cb49ca316fe13e3e70bd19e7ab0863e50/src-built-in/preloads/zoom.js#L180-L183) for definitions.

This PR adds a component config setting:
`foreign.components.['Window Manager'].zoomDefault`

**Comment by JMCCROWELL** [comment]
@kriswest: I think that the Config Reference makes the most sense in this instance. Good question, thanks!

**Comment by kriswest** [comment]
@JMCCROWELL I wrote up the docs changes and put them on finsemble PR: https://github.com/ChartIQ/finsemble/pull/1801

**Pull request merged**